src/surroundings_menu.cpp

```diff
@@ -0,0 +1,1160 @@
+            if( !monster_data.selected_entry ||
+                std::find( monster_data.filtered_list.begin(), monster_data.filtered_list.end(),
+                           monster_data.selected_entry ) == monster_data.filtered_list.end() ) {
+                monster_data.selected_entry = monster_data.filtered_list.front();
+            }
+            std::string last_category;
+            int entry_no = 0;
+            for( map_entity_stack<Creature> *it : monster_data.filtered_list ) {
+                if( monster_data.draw_categories ) {
+                    draw_category_separator( it->get_category(), last_category, 1 );
+                }
+                bool is_selected = it == monster_data.selected_entry;
+                ImGui::TableNextColumn();
+                ImGui::PushID( entry_no++ );
+                ImGui::Selectable( "", &is_selected,
+                                   ImGuiSelectableFlags_AllowItemOverlap | ImGuiSelectableFlags_SpanAllColumns, ImVec2( 0, 0 ) );
+                if( is_selected ) {
+                    if( auto_scroll ) {
+                        ImGui::SetScrollHereY();
+                        auto_scroll = false;
+                    }
+                    monster_data.selected_entry = it;
+                }
+                ImGui::SameLine( 0, 0 );
+                ImGui::PopID();
+                const Creature *mon = it->get_selected_entity();
+                const Character &you = get_player_character();
+                const bool inattentive = you.has_trait( trait_INATTENTIVE );
+                bool sees = mon->sees( you ) && !inattentive;
+                cataimgui::draw_colored_text( sees ? "!" : " ", c_yellow );
+
+                ImGui::TableNextColumn();
+                std::string mon_name;
+                if( mon->is_monster() ) {
+                    mon_name = mon->as_monster()->name();
+                } else {
+                    mon_name = mon->disp_name();
+                }
+                nc_color mon_color = mon->basic_symbol_color();
+                cataimgui::draw_colored_text( mon_name, mon_color );
+                if( is_selected ) {
+                    ImGui::SetScrollHereY();
+                }
+
+                ImGui::TableNextColumn();
+                nc_color hp_color = c_white;
+                std::string hp_text;
+
+                if( mon->is_monster() ) {
+                    mon->as_monster()->get_HP_Bar( hp_color, hp_text );
+                } else {
+                    std::tie( hp_text, hp_color ) = get_hp_bar( mon->as_npc()->hp_percentage(), 100 );
+                }
+                const int bar_max_width = 5;
+                const int bar_width = utf8_width( hp_text );
+                if( bar_width < bar_max_width ) {
+                    hp_text += "<color_c_white>";
+                    for( int i = bar_max_width - bar_width; i > 0; i-- ) {
+                        hp_text += ".";
```

Comment:
Better to just call `ImGui::TextUnformatted` in a loop than to repeatedly append to a string.